src/main/java/org/zalando/nakadi/repository/kafka/KafkaLocationManager.java

```diff
@@ -28,13 +31,31 @@
     private final Properties kafkaProperties;
     private final KafkaSettings kafkaSettings;
     private final ScheduledExecutorService scheduledExecutor;
+    private final Set<Runnable> ipAddressChangeListeners;
 
     public KafkaLocationManager(final ZooKeeperHolder zkFactory, final KafkaSettings kafkaSettings) {
         this.zkFactory = zkFactory;
         this.kafkaProperties = new Properties();
         this.kafkaSettings = kafkaSettings;
+        this.ipAddressChangeListeners = new HashSet<>();
+        this.updateBootstrapServers();
+        this.subscribeForBrokersChange(zkFactory);
         this.scheduledExecutor = Executors.newSingleThreadScheduledExecutor();
-        this.scheduledExecutor.scheduleAtFixedRate(this::updateBootstrapServers, 0, 1, TimeUnit.MINUTES);
+        this.scheduledExecutor.scheduleAtFixedRate(this::updateBootstrapServers, 1, 1, TimeUnit.MINUTES);
+    }
+
+    private void subscribeForBrokersChange(final ZooKeeperHolder zkFactory) {
+        try {
+            zkFactory.get().getChildren()
+                    .usingWatcher((Watcher) event -> {
+                        this.ipAddressChangeListeners.forEach(Runnable::run);
+                        this.scheduledExecutor.schedule(this::updateBootstrapServers, 0, TimeUnit.MILLISECONDS);
+                        this.scheduledExecutor.schedule(
+                                () -> subscribeForBrokersChange(zkFactory), 0, TimeUnit.MILLISECONDS);
```

Comment:
With this approach you may loose several notifications, it is better to create watcher while getting nodes.
And, in case if it is initialization or reaction on watcher - recreate watcher.